tensorflow_federated/python/research/robust_aggregation/robust_aggregation_test.py

```diff
@@ -0,0 +1,233 @@
+
+  return model_fn
+
+
+class DummyClientComputation(optimizer_utils.ClientDeltaFn):
+  """Client TensorFlow logic for example.
+  Designed to mimic the class `ClientFedAvg` from federated_averaging.py
+  """
+
+  def __init__(self, model, client_weight_fn=None):
+    """Creates the client computation for Federated Averaging.
+    Args:
+      model: A `tff.learning.TrainableModel`.
+      client_weight_fn: Optional argument is ignored
+    """
+    del client_weight_fn
+    self._model = model_utils.enhance(model)
+    py_typecheck.check_type(self._model, model_utils.EnhancedTrainableModel)
+    self._client_weight_fn = None
+
+  @property
+  def variables(self):
+    return []
+
+  @tf.function
+  def __call__(self, dataset, initial_weights):
+    del initial_weights
+    model = self._model
+
+    @tf.function
+    def reduce_fn_num_examples(num_examples_sum, batch):
+      """Count number of examples"""
+      num_examples_in_batch = tf.shape(batch['x'])[0]
+      return num_examples_sum + num_examples_in_batch
+
+    @tf.function
+    def reduce_fn_dataset_mean(sum_vector, batch):
+      """Sum all the examples in the local dataset"""
+      sum_batch = tf.reshape(tf.reduce_sum(batch['x'], [0]), (-1, 1))
+      return sum_vector + sum_batch
+
+    num_examples_sum = dataset.reduce(
+        initial_state=tf.constant(0), reduce_func=reduce_fn_num_examples)
+
+    example_vector_sum = dataset.reduce(
+        initial_state=tf.zeros((dim, 1)), reduce_func=reduce_fn_dataset_mean)
+
+    # create an ordered dictionary with the same type as model.trainable
+    # containing a mean of all the examples in the local dataset
+    # Note: this works for a linear model only (as in the example above)
+    key = list(model.weights.trainable.keys())[0]
+    weights_delta = collections.OrderedDict(
+        {key: example_vector_sum / tf.cast(num_examples_sum, tf.float32)})
+
+    aggregated_outputs = model.report_local_outputs()
+    weights_delta, has_non_finite_delta = (
+        tensor_utils.zero_all_if_any_non_finite(weights_delta))
+
+    weights_delta_weight = tf.cast(num_examples_sum, tf.float32)
+
+    return optimizer_utils.ClientOutput(
+        weights_delta, weights_delta_weight, aggregated_outputs,
+        tensor_utils.to_odict({
+            'num_examples': num_examples_sum,
+            'has_non_finite_delta': has_non_finite_delta
+        }))
+
+
+def build_federated_process_for_test(model_fn, num_passes=5, tolerance=1e-6):
+  """ Analogue of `build_federated_averaging_process`, but with
+  client_fed_avg replaced by the dummy mean computation defined above.
+  """
+
+  server_optimizer_fn = lambda: tf.keras.optimizers.SGD(learning_rate=1.0)
+
+  def client_fed_avg(model_fn):
+    return DummyClientComputation(model_fn(), client_weight_fn=None)
+
+  # Build robust aggregation function
+  with tf.Graph().as_default():
+    # workaround since keras automatically appends "_n" to the nth call of `model_fn`
```

Comment:
```suggestion
    # workaround since keras automatically appends "_n" to the nth call of
    # `model_fn`
```

TFF  enforces 80-character line width limits.